swift/ql/lib/codeql/swift/dataflow/internal/DataFlowPrivate.qll

```diff
@@ -785,6 +827,156 @@ private module OutNodes {
 
 import OutNodes
 
+/**
+ * Holds if there is a data flow step from `e1` to `e2` that only steps from
+ * child to parent in the AST.
+ */
+private predicate simpleAstFlowStep(Expr e1, Expr e2) {
+  e2.(IfExpr).getBranch(_) = e1
+  or
+  e2.(AssignExpr).getSource() = e1
+  or
+  e2.(ArrayExpr).getAnElement() = e1
+}
+
+private predicate closureFlowStep(CaptureInput::Expr e1, CaptureInput::Expr e2) {
+  simpleAstFlowStep(e1, e2)
+  or
+  exists(Ssa::WriteDefinition def |
+    def.getARead().getNode().asAstNode() = e2 and
+    def.assigns(any(CfgNode cfg | cfg.getNode().asAstNode() = e1))
+  )
+  or
+  e2.(Pattern).getImmediateMatchingExpr() = e1
```

Comment:
Note that `e2` is of type `CaptureInput::Expr`, and `CaptureInput::Expr` is actually an [AstNode](https://github.com/github/codeql/pull/14577/files/862de152a1308d02ee43fa01f3cf764357303f7f#diff-514dd0988f7b47e40ff8c1e66ea886d1f947d047da108270ed0ddec75c2856c1R889).

Reply:
Yes, that's why compilation doesn't fail. I'm asking whether `Expr` and `Pattern` might be sufficiently disjoint (even though both are subtypes of `AstNode`) that this will block flow?

Reply:
I think I'll leave this to another PR. I've discovered that the `WriteDefinition::assigns` predicate doesn't really do the right thing at all. For a variable declaration such as:
```
var x = 0
```
the `WriteDefinition::assigns(value)` predicate holds where `value` is the left-hand side of the declaration (i.e., the lvalue of `x`) instead of the right-hand side (i.e., `0`). Once that's fixed (which I'll do in a later PR) we can remove this disjunct and have `CaptureInput::Expr` be `Expr`.